Digital ocean saltcloud dns record creation and deletion implemented

aphor opened the pull request:
Due to design problems in salt-cloud (no way to make configuration data available to destroy methods), DNS record deletion unconditionally deletes all DNS records matching the given FQDN if it can.

Commit:
Revoke auth immediately on minion key deletion

Fixes #25124

ticosax committed:
docker.version_info is now provided.

ticosax committed:
Do not pass any argument to docker.start

docker.create is responsible to manage all of them, since docker API 1.16

ticosax committed:
prevent potential error while reporting mismatch versions to user.

ticosax committed:
fixup! Do not pass any argument to docker.start

Commit:
Set default

lathama committed:
Sneaky white space

Using the web based editor and some white space found its way in.

Commit:
Use eipalloc instead of eni in EC2 interface properties example

Fixes #27815

cro committed:
WIP modules and states for managing Dell FX2 chassis via salt-proxy

eliasp committed:
Provide all git pillar dirs in `opts[pillar_roots]`

Do not only provide the `pillar_roots` directory of the currently
evaluated pillar SLS.
This makes it impossible to use `include` for pillar SLS located in
another git repository.

Fixes #27932

Commit:
Clarify that apache2-utils is for Debian-based distros

Commit:
Clarify that apache2-utils is for Debian-based distros

Commit:
Changed grain to look at osmajorrelease instead of osrelease

The-Loeki committed:
Fix CloudStack cloud for new 'driver' syntax

The-Loeki committed:
Pop deprecated 'provider' into new 'driver' key

Even that I can do @basepi :D

twangboy committed:
Added new functionality

Added the following:
- task.info
- task.status
- task.stop

twangboy committed:
Clarified an error

techhat committed:
Change __opts__ to self.opts

twangboy committed:
Added run_wait function

jeffreyctang committed:
fix parse of { on next line

jeffreyctang committed:
lint fixes

Commit:
Decode unicode names in fileclient/server

Closes #27712

terminalmage committed:
Add note about dockerng.inspect_image usage

Should clear up apparent confusion raised in
https://github.com/saltstack/salt/issues/27976

Commit:
Fix regression with key whitespace

Port #22115 into 2015.8.

Closes #27750

Commit:
Fixed trailing whitespace lint

Commit:
Add example for salt-proxy over SSH

Commit:
Pip state run result should be False, not None, if installation error occurs.

Fixes #27971

Commit:
Versionadded for clean_file option for pkgrepo

Refs #19561

jacobhammons committed:
Changed current release to 8.1 and added CVEs to release notes

twangboy committed:
Fix for #20678

win_pkg.py
- Added the option to use the windows task scheduler (`use_scheduler`)
- Applied to both pkg.install and pkg.remove

win_task.py
- Added new parameter `force` to task.create
- Forces the creation of the update of the task if it already exists
- Fixed problem with `task.run_wait` sometimes erroring when the task finished
- Seperated out the cmd and its arguments, they must be explicitely passed,
`task.add_action` will not try to parse the cmd parameter for arguments

win_service.py
- Added new parameters for win_task.py
- Explicitely send command and arguments
- Add `force` parameter to create the task everytime

cro committed:
Lint fixes and some changes by @rallytime

Commit:
Make load beacon config into list

Closes #27949

The load beacon config was incorrectly formatted as a list when it should have been a dict.

@jacobhammons this will require an entry in the release notes

cro committed:
Cleanup, add blade_idrac stub

jacobhammons committed:
added link to grains security FAQ to targeting and pillar topics.

Commit:
Don't fail if pip package is already present and pip1 is installed

Refs #21845

Commit:
Remove note about pip bug with pip v1 vs pip v2 return codes

Refs #27996

Commit:
Maintain stateful output if something went wrong running the pip command

Commit:
Cleanup pkgin isatty mess

Commit:
Provide empty string as default stdout instead of None

twangboy committed:
Fix PR #26336

Problem with the return value from the registry lookup.

twangboy committed:
Fix PR #26336

Problem with the return value from the registry lookup.

Commit:
Clean up stack trace when something goes wrong with minion output

Fixes #28010

Commit:
Add example to Writing Grains of how grains can be loaded twice

Fixes #27938

Commit:
Clarify loading vs rendering the final grains data structure

Commit:
Clean up version added and deprecated msgs to be accurate

Fixes #27828

Commit:
LazyLoader deepcopy fix.

Don't keep module object as a field of LazyLoader that breaks
deepcopying.

jacksontj committed:
Correctly pop the future off the future_map

jacobhammons committed:
Updated Debian and Ubuntu repo paths with new structure for 2015.8.1

ryancurrah committed:
Update chocolatey.py

Change 'install' and 'uninstall' state names to the more state-full naming convention 'installed' and 'uninstalled'.

twangboy committed:
Fixed win_path.py

Moved vdata and split to the same line
No need to check for string types

cro committed:
Make SSH connection 'persistent'.  Note that right now this requires 'multiprocessing: False' in /etc/salt/proxy.

jacobhammons committed:
Updated release notes with additional CVE information

twangboy committed:
Fixed win_useradd.py

The real fix. Instead of getting vdata in the calling function,
_get_userprofile_from_registry should return just the profile and not a dict
full of a bunch of crap. This also fixes the debug log for the
_get_userprofile_from_registry function.

twangboy committed:
Fixed win_useradd.py

The real fix. Instead of getting vdata in the calling function,
_get_userprofile_from_registry should return just the profile and not a dict
full of a bunch of crap. This also fixes the debug log for the
_get_userprofile_from_registry function.

cro committed:
Lint, remove debug.

cro committed:
Default multiprocessing to False since anything that needs salt.vt will have trouble with our forking model.

cro committed:
Spelling.

Commit:
Don't deepcopy lazy loader just to iterate over keys.

erchn committed:
Rackspace support for switft module.

Allow setting auth_version, also don't make `password` a required field, can also use api_key.
Pass api_key in kwargs as `key`, that's what's expected in the utils module.
Also update notes with those details and a recommendation for using version 1 auth.

ryancurrah committed:
Update chocolatey.py

Added deprecation path support and docs as per jfindlay

erchn committed:
Allow passing in auth_version, defaulting to 2.

erchn committed:
Allow passing in auth_version, defaulting to 2.

erchn committed:
cleanup whitespace, default to None to be consistent with profile

erchn committed:
remove extra spaces

cro committed:
Spelling, lint.

erchn committed:
default auth_version = 2

erchn committed:
Revert "Allow passing in auth_version, defaulting to 2."

This reverts commit 31ff2abd03ee9ba0238f5e1c12a398519e06b140.

erchn committed:
Revert "Allow passing in auth_version, defaulting to 2."

This reverts commit 31ff2abd03ee9ba0238f5e1c12a398519e06b140.

gtmanfred committed:
use the correct discover_extensions

This shouldn't have ever been used, it should have always used the real
discover_extension in the main client module.

Right now nova is broken with the current novaclient release, which
removes _discover_extensions from OpenStackComputeShell

ryancurrah committed:
Update chocolatey.py

As per linter incompatible Python 3 code found. Applying proposed fix.

jfindlay committed:
fix repo path in ubuntu installation documentation

Fixes #28038.

gtmanfred committed:
the nova driver does not require libcloud

The only thing used from libcloudfuncs is the script and reboot, which
both hook through the regular connection, and are namespaced down to the
nova driver.  they use the regular conn, which hooks to the SaltNova
library in order to reboot the server.

Commit:
Add pkg install and remove functions

Commit:
Restore FTP functionality to file client

Closes #27534

cxxcjeremy committed:
Implementation for "create_record" DigitalOcean DNS support in salt-cloud

cxxcjeremy committed:
Implementation for "create_record" DigitalOcean DNS support in salt-cloud

cro committed:
Documentation.

cro committed:
More Documentation.

Commit:
Use a better method for checking dynamodb table existence

Commit:
Purge iam user attributes before deleting iam user

It's necessary to remove a number of iam user attributes before
deleting the user, or the user can't be deleted. This change
purges those attributes by default.

Commit:
Lint

Commit:
Context manager

SmithSamuelM committed:
fixed unittests missing close of roadstack caused error on other tests

Commit:
Fix #28060

Correct the `pip` command. install should be getting `global-options` not `pip` itself.

SmithSamuelM committed:
Raet Salt broken when config moved to package directory
The path to the master.flo file no longer worked This fixes

SmithSamuelM committed:
Fixed exception in loader when no file extension

SmithSamuelM committed:
Changed reference to reflect refactor of ioflo package locations as of ioflo 1.2.3
Deprecated package locations still supported in ioflo for now

SmithSamuelM committed:
Changed reference to reflect refactor of ioflo package locations as of ioflo 1.2.3
Deprecated package locations still supported in ioflo for now

jacksontj committed:
Use the generic `text` attribute, not .body of the handler

fixes #27794 until we can simplify this somewhat crazy HTTP lib :/

s0undt3ch committed:
Don't install futures under Python 3 and above

aboe76 committed:
update suse spec file to 2015.8.1

blueyed committed:
dockerng: use error from modules.dockerng in states' __virtual__

This improves the return value for errors in __virtual__ by using the
information from `modules.dockerng`.

Ref: https://github.com/saltstack/salt/issues/27643#issuecomment-148942821

lomeroe committed:
adding convert_cidr function to convert a cidr to a network and netmask

lomeroe committed:
correcting versionadded statement

Commit:
Fix lint errors

Commit:
Replace option 'i' with an explicit queryformat

Option 'i' uses a predefined queryformat that is not necessarily compatible with the code used to parse it.

For example: an exception was being thrown on my CentOS 6.7 machines because the default queryformat was placing multiple keys on a signle line.

I have therefore replaced the 'i' option with a hardcoded queryformat to ensure consistancy across all systems.

Commit:
Don't deepcopy just to iterate

Commit:
Fix lint errors

topochan committed:
Fix Doc in jinja template, 

In the example expr_form is using pillar instead of grain when in the doc is explained the use of a grain for the role

Commit:
Fix for trace dump on failing imports for win32com & pythoncom 4 win_task

Commit:
fix for glance state trace error on import failure

Commit:
Lint

Commit:
Revert "Update __init__.py"

This reverts commit 5664de6610de024e09b35d919a10c715370bd539.

The commit introduced a specific implementation of WheelClient.cmd()
that calls low(fun, low) method passing args as low argument that's
wrong, args and kwargs should be packed into low that should be a dict.

lomeroe committed:
passing cidr parameter through calc_net to verify we have the network address of the cidr

Commit:
Minor optimization for cloud cache

Closes #15741

lomeroe committed:
correction to versionadded documentation

Commit:
Addition of SysFS module

Commit:
Add clone for qemu

Commit:
Don't revoke auth if no aes key rotate

jacksontj committed:
Revert "Use the generic `text` attribute, not .body of the handler"

This reverts commit b1bf79821d2d9ae3fb1a02abfa859445f335bb49.

jacksontj committed:
Switch fileclient to not stream at all.

This uses the `text` return-- which is a bit of a misgnomer since none of the implementations make sure its text-- they just dump the body in (except requests-- which now does as well).

This is hopefully one of the last patches before we clean this all up :)

carlpett committed:
win_firewall: Report correct result in test mode

jacksontj committed:
For all multi-part messages, check the headers. If the header is not your minion-id or a broadcast, drop the message.

Fix for #26844

gtmanfred committed:
add option to allow slack returner to show changes

I don't need all the extra information in slack, I only need to know
what changes or what failed.

ajacoutot committed:
OpenBSD salt package: update list of dependencies.

s0undt3ch committed:
Allow providing a timeout when attempting to connect to a master

s0undt3ch committed:
Return the added process

Commit:
Add documentation for proxy minion ssh

Commit:
Don't rotate key if option to not rotate is set

jfindlay committed:
add reason for not loading localemod

Fixes #28080.

cxxcjeremy committed:
implement DigitalOcean salt-cloud vm options create_dns_record, dns_domainname, and dns_hostname

cxxcjeremy committed:
implement DigitalOcean salt-cloud vm options create_dns_record, dns_domainname, and dns_hostname

gtmanfred committed:
allow slack to be formatted in yaml

skizunov committed:
'RuntimeError: IOLoop is closing' thrown in Minion when using TCP transport

This issue was caused by the fix to Issue #26889. That fix caused the
io_loop to close in: 'salt.utils.async.SyncWrapper.__del__' before
closing the stream in 'salt.transport.SaltMessageClient.destroy'. The
side effect of this was that in tornado 4.2.1, tornado/ioloop.py lines
911-914:
    def add_callback(self, callback, *args, **kwargs):
        with self._callback_lock:
            if self._closing:
                raise RuntimeError("IOLoop is closing")

This RuntimeError exception would be raised. The reason why it is
raised is that the read future is still pending, and so it will invoke
set_exception(StreamClosedError()) on that future when closing the
stream. That set_exception will try to invoke a callback and run into
the code above which would throw the RuntimeError because the io_loop
is closed. The work-around is to mark the stream's '_read_future' as
None to prevent it from calling set_exception on it.

A better long term solution would be to rearchitect SaltMessageClient
to no longer use 'read_until' nor 'read_bytes' and instead use
'read_until_close' since 'read_until_close' is designed to have a read
active at the time of close whereas the others are not. This would
allow the removal of all the work-arounds in
'SaltMessageClient.destroy'.

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

Commit:
Set logfile permsissions correctly

Closes #28074

cro committed:
More docs.

Commit:
Add created reactor event to lxc cloud driver

Fixes #27940

cro committed:
Lint

terminalmage committed:
Add explanation of file_client: local setting masterless mode

cro committed:
Lint

cro committed:
Lint

lomeroe committed:
switching to use salt.ext.ipaddress module - thanks for the suggestion @The-Loeki

lomeroe committed:
correcting return statements from ipaddress object

cro committed:
Lint

skrobul committed:
s6 service module

Commit:
Clean up stacktrace when master can't be reached in lxc cloud driver

Fixes #23655

lomeroe committed:
lint fix

Commit:
Remove unnecessary padding from rpm.info

cxxcjeremy committed:
Implement (broken) delete_dns_record support.. cloud config is weird, still needs help.

Commit:
Correction pylint error

alexproca committed:
Import public key

Commit:
Correct plynt

roxit committed:
Make a copy of the merge argument in grains.filter_by

Multiple calls of grains.filter_by in the same session might change
the merge argument, which is sometimes __pillar__, resulting in
incorrect context in map.jinja.

This change make a copy of the merge argument before passing it to
dictupdate.update.

Closes: #28127

bogdanr committed:
Added a bunch of instance sizes and updated some outdated ones

bogdanr committed:
Documented import_keypair for the ec2 driver

Commit:
Set a fallback HWM

This mostly happens in the test suite when opts are setup by hand, but a safe default is a good idea regardless.

s0undt3ch committed:
Remove the TODO, the function should not be removed.

s0undt3ch committed:
Py3 compat

alexproca committed:
Added missing comma

s0undt3ch committed:
Don't use the deprecated `imp.get_suffixes()` under Py3.

s0undt3ch committed:
Py3 compat

bernieke committed:
fix unicode pillar values #3436

Commit:
Pylint fix

Commit:
Versionadded for new function

Commit:
Don't use RAM for root disk size in linode.py

Fixes #28118

Commit:
utilize salt.utils.fopen() in SysFS write function

Commit:
SysFS module small default val fix

Commit:
Removed blank line at end of chassis.py

Commit:
removed extraneous file

Commit:
Lint dracr.py

Commit:
Add clarification to cloud profile doc about host

Commit:
Improve documentation of `check_cmd`.

* Make the distinction between requisite `check_cmd` and `file.managed`
  `check_cmd`.

* Be more clear about `file.managed` `check_cmd`, the temporary file,
  and when the new contents are written or discarded.

* In requisite `check_cmd` use a concise `!` rather than the awkward
  `&& exit 1 || exit 0`.

* A few minor, stylistic changes.

lorengordon committed:
Add `defaults` and `context` parameters to cmd.script state

Fixes #9153

Commit:
Update doc

lorengordon committed:
Add type-checking for `defaults` and `context` in cmd.script

cxxcjeremy committed:
DNS record creation and deletion now works, but due to salt-cloud config design flaws, deletion is unconditional not configurable

cxxcjeremy committed:
DNS record creation and deletion now works, but due to salt-cloud config design flaws, deletion is unconditional not configurable

Commit:
philips_hue fails during unit tests.

It appears that it cannot index 'proxymodule' and doesn't properly handle the
exception.  This may not be the best way, but it will properly skip
initialization if settings are unavailable.

basepi committed:
Quiet pylint

rallytime commented:
Thanks for submitting this @aphor. There are quite a few [lint errors](https://jenkins.saltstack.com/job/salt-pr-lint-n/10205/violations/file/salt/cloud/clouds/digital_ocean.py/). Can you clean those up?

Also please note that we need to support Python 2.6, so doing things like `something {}'.format(somevar)` doesn't work in 2.6, so if you could clean those up as well, that'd be great.

I've also asked @techhat to look this over as well.

rallytime commented on salt/cloud/clouds/digital_ocean.py:
```diff
@@ -720,20 +772,18 @@ def destroy(name, call=None):
 
 
-def create_dns_record(hostname, ip_address):
+def post_dns_record(dns_domain, name, record_type, record_data):
```
If you rename this function and change it's API signature, you'll need to put this on a deprecation path and make sure this is backwards compatible until the deprecation is officially removed. 

Docs on how to do this can be found here: https://docs.saltstack.com/en/latest/topics/development/deprecations.html

One thing to consider: Does this function name need to actually change here? Can you keep the name and required args and then add additional args as kwargs to not affect the API? We like to keep API changes to a minimum. Obviously there are times when deprecations are necessary, but they can be a hassle if not absolutely needed.

aphor replied:
Thanks for the onboarding guidance! I'll add the deprecation warnings.

I think this needs to change. A) create_dns_record() as used should be deprecated. I'll try to justify that first. B) it's not clear that this function should be an API function, and I suggest it probably should not be. I'll try to justify that second.

There are lots of DNS records that have nothing to do with hostnames: SOA, NS, TXT, SRV for example. It would be wrongheaded to design the API to prevent their use. There are at least two types of DNS records that need to be unambiguously distinct: A and AAAA.

In the case of DigitalOcean, their REST interface is a POST HTTP method. We're wrapping that, and calling it "create" is cargo-cultish, and I'm changing the arguments, so presumably we *could* have both. IMHO, at the very least create_dns_record() should be changed create_dns_A_record() if you're into Knuth-y literate programming.

Also 'create_dns_record' is a nice declarative to use in the context of a VM config. There is a convention in the cloud drivers to represent those as identical variable identifiers. That would preclude using the configuration item in the same variable scope as the function that implements it: ouch. As far as API compatibility: I don't think anything outside of the DigitalOcean driver code calls this. AFAICT the upstream exposure is through salt-cloud config options.

rallytime replied:
That's all fine. Just figured I would point it out. As long as the correct deprecation warnings are in place, that is fine to change.

rallytime commented on salt/cloud/clouds/digital_ocean.py:
```diff
@@ -741,23 +791,31 @@ def create_dns_record(hostname, ip_address):
 
 
-def delete_dns_record(hostname):
+def destroy_dns_records(fqdn):
```
Same comment about deprecations applies here as well.

basepi committed:
remove pylint disables, as it's broken in develop anyway

(plus the pylint disables didn't work)

Commit:
Always get default option settings from salt.config

It is a poor practice to duplicate settings and scatter them through multiple
files.  Configuration options have been centralize in salt.config, yet
salt.utils.event repeats many of those settings for the benefit of unit tests
that pass only partial "opts" parameters.  Rather than change all of the unit
tests (and other callers) to provide sufficient settings in the "opts"
parameters, the various classes that need opts should start with the default
options provided by salt.config as a base and then the passed-in "opts"
parameter to __init__() functions should override the base options.

  * Use the appropriate salt.config.DEFAULT_MASTER_OPTS or
    salt.config.DEFAULT_MINION_OPTS.

  * Remove all of the spurrious, redundant default settings from
    salt.utils.event.

  * Drop use of "opts.get(KEY)" in favor of "opts[KEY]" - KeyError exceptions
    are a good thing because they indicate that an appropriate default is not
    in salt.config.

TODO: Perform similar clean-ups through other files in the code base.

Commit:
Add failed minion information to summary (--summary argument) information:

  * Add count of "# of minions with errors"
  * Add specific failed minion IDs when --verbose is used

aphor commented:
Python 2.6 supported? OK I can understand supporting 2.6 minions, but for code that doesn't run in the minion? Python 2.6.x isn't even getting security patches any more since 2013. Should I open an issue for this?

https://www.python.org/download/releases/2.6.7/:  
> Python 2.6 is now in security-fix-only mode; no new features are being added, and no new bug fix releases are planned. We intend to provide source-only security fixes for the Python 2.6 series until October 2013 (five years after the 2.6 final release). For ongoing maintenance releases, please see the Python 2.7 series. The NEWS file lists every change in each alpha, beta, and release candidate of Python 2.6.

Commit:
Additional corrections to use option defaults directly from salt.config

  * correct 'id' entry in salt.config
  * add 'event_match_type' in salt.config
  * ensure that hash is only calculated in minion since it isn't needed by the master

rallytime commented:
Thanks for the questions/comments. You don't need to open a new issue about Python 2.6 support. We must support Python 2.6 because you can still run a Master and a Minion on distributions that are not EOL yet and we support here at SaltStack. For example, CentOS 5 and 6 are still being maintained, and we are still building packages for those distros, so Python 2.6 compatibility must be maintained for both masters and minions.

In terms of the code not running on the minion: salt-cloud is it's own process and can be run from a master or a minion. Most people run it from a master's perspective/approach, but there are many others who run it from a masterless infrastructure, too.

cxxcjeremy committed:
pylint: that's not attractive. me: Cleanup done\!

cxxcjeremy committed:
pylint: that's not attractive. me: Cleanup done\!

cxxcjeremy committed:
put delete_dns_record() and create_dns_record() back, with deprecation warnings

cxxcjeremy committed:
put delete_dns_record() and create_dns_record() back, with deprecation warnings

cxxcjeremy committed:
lint: indentation fix

cxxcjeremy committed:
lint: indentation fix

cxxcjeremy committed:
lint: PEP8E302 two blank likes around top level function defs fixed

cxxcjeremy committed:
lint: PEP8E302 two blank likes around top level function defs fixed

cxxcjeremy committed:
lint: misc lint

cxxcjeremy committed:
lint: misc lint

aphor commented on salt/cloud/clouds/digital_ocean.py:
```diff
@@ -414,12 +447,23 @@ def __query_node_data(vm_name):
+     
+    # add DNS records, set ssh_host, default to first found IP, preferring IPv4 for ssh bootstrap script target
+    addr_families, dns_arec_types = (('v4', 'v6'), ('A', 'AAAA'))
+    arec_map = dict(zip(addr_families, dns_arec_types))
```
The lint checker Jenkins is running says this is incompatible python 3 code, but I don't think so.

Python 2.7.5 (default, Jun 24 2015, 00:41:19) 
[GCC 4.8.3 20140911 (Red Hat 4.8.3-9)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> addr_families, dns_arec_types = (('v4', 'v6'), ('A', 'AAAA'))
>>> arec_map = dict(zip(addr_families, dns_arec_types))
>>> arec_map
{'v4': 'A', 'v6': 'AAAA'}

aphor commented:
I added the deprecation stuff to ease the changes to the driver functions. I de-linted what makes sense. There are still lint errors. I rebased from develop, and develop brought more errors that aren't mine. I think I'm mostly done with this pull request unless there's more work from review. I'll add some documentation.

cxxcjeremy committed:
Change documentation to reflect actually implemented behavior

aphor commented:
I guess I should babysit this and keep rebasing until all the junk I pulled in from the last rebase stops failing tests?

rallytime commented:
@aphor Sorry I didn't get back to you yet. There are certainly some problems with the develop branch that we're working on. I'll give this another thorough review next week.

In the mean time, there are still some [outstanding pylint errors](https://jenkins.saltstack.com/job/salt-pr-lint-n/10285/violations/file/salt/cloud/clouds/digital_ocean.py/) related to your change that need to be addressed. The python three incompatibilities give hints for how to fix them in the [console output](https://jenkins.saltstack.com/job/salt-pr-lint-n/10285/console) of the pylint run. Look for "Proposed fix" in the output and it will show you a rudimentary diff out how to fix it.

rallytime commented:
Refs #28107.

rallytime commented on salt/cloud/clouds/digital_ocean.py:
```diff
@@ -694,17 +738,25 @@ def destroy(name, call=None):
+    delete_dns_record = True
 
-    if delete_record and not isinstance(delete_record, bool):
+    if delete_dns_record and not isinstance(delete_dns_record, bool):
```
If you're setting `delete_dns_record = True` above, then this check isn't even necessary.

aphor replied:
The explanation/fix for this also applies to the next comment.

rallytime commented on salt/cloud/clouds/digital_ocean.py:
```diff
@@ -694,17 +738,25 @@ def destroy(name, call=None):
 
-    if delete_record:
-        delete_dns_record(name)
+    if delete_dns_record:
```
These `if`/`else` statements also seem to be confusing since you've already set the value to `True` above. What is the intention here?

aphor replied:
I should add comments. I really struggled to find a way to get configuration data so that delete_record can be asserted by configuration instead of code, but I don't see a good way, and implementing this should be a separate issue/enhancement/bug.

techhat replied:
If you don't add comments, then it won't be long before somebody else removes what seems to be superfluous code.

rallytime replied:
Ok. If you add some comments about that and be explicit about what is happening here, then that is fine. Please add comments to the other place where you said "The explanation/fix for this also applies to the next comment.".

aphor replied:
I'll add the comments to the part where I hard-set the delete-dns-record = True, and then if the conditionals get deleted as superfluous, it's not a big loss, just kinda sad we couldn't fix the lack of config data access to make the behavior configurable.

rallytime replied:
Oh, I actually added some here: https://github.com/saltstack/salt/pull/28369/files#diff-6e82f218b20dac0495868c38492c4c49R763 

I made this PR to get those pylint errors fixed and also made some comments about what was going on there. I think the previous comments that you added with the TODO and those additions will be fine.

rallytime commented on salt/cloud/clouds/digital_ocean.py:
```diff
@@ -720,12 +772,38 @@ def destroy(name, call=None):
+    return __deprecated_create_dns_record(hostname, ip_address)
+
+
+def __deprecated_create_dns_record(hostname, ip_address):
```
Nice! These deprecations look great. Thanks for doing those.

rallytime commented on salt/cloud/clouds/digital_ocean.py:
```diff
@@ -414,12 +447,23 @@ def __query_node_data(vm_name):
+                vm_['ssh_host'] = ip_address
+    
     vm_['key_filename'] = key_filename
-    vm_['ssh_host'] = ip_address
```
`vm_['ssh_host']` must be set before calling the `salt.cloud.bootstrap` function (next line after this one) as we need this value [here](https://github.com/saltstack/salt/blob/develop/salt/utils/cloud.py#L392). Can you rewrite this part so that this line won't overwrite your `for` loop changes in lines 454-465 but that we can also maintain current functionality?

aphor replied:
In most cases, the salt master (or whatever host is running the bootstrap), will only be able to reach the public IP of the new DigitalOcean droplet. AFAIK, *all* DigitalOcean droplets have a public IP. If somehow DigitalOcean fails to provide (or indicate) a public interface IP address, then bootstrap will fail on SSH connection failure anyway. I propose to add a warning message in that (corner) case. The current functionality will crash with an exception "NameError: name 'ip_address' is not defined" in that case. Do we need to preserve that failure mode?

rallytime replied:
Ok, so the expectation that the `vm_['ssh_host'] = ip_address` line in 464 will likely be hit, correct? I just don't want to see a `KeyError` pop up when we get the the call to the bootstrap script. So a warning message and an exit before or try/except around the bootstrap call would be great.

rallytime commented:
@aphor I've left a couple of new comments about your changes. Overall this looks great! However those remaining issues will be cleaned up before merging this in. Please let me know if you have any questions! :)

cxxcjeremy committed:
be explicit about failures to find an IP for ssh bootstrapping and about conditional code to delete DNS records being unconditional

rallytime commented on salt/cloud/clouds/digital_ocean.py:
```diff
@@ -694,17 +742,28 @@ def destroy(name, call=None):
+    #         delete them if we can.
+    delete_dns_record = True
+
+    if delete_dns_record and not isinstance(delete_dns_record, bool):
```
This should be simplified to use `if not isinstance(delete_dns_record, bool):`` since `delete_dns_record` is already True.

rallytime commented:
Thanks for those changes! I've grabbed the pylint errors in the PR referenced above. When that is merged in, your commits will be pulled in, too. Thanks again for your work here!

aphor commented:
rebased again. it was a PAIN because develop had reverted a bunch of stuff that I picked up in an earlier rebase. I had to force the push. I think I'm misunderstanding what GitHub behind the scenes is doing with pull requests. For conflicts, the only files the merge should take from mine are:
- salt/cloud/clouds/digital_ocean.py
- doc/topics/cloud/digitalocean.rst

rallytime commented:
The relevant commits have been merged into develop with Pull Request #28369. Your commits are all intact there, though. As such, I am going to close this PR. 

Thanks again for your work here!
